llvm/lib/IR/ConstantRange.cpp

```diff
@@ -1520,15 +1520,72 @@ ConstantRange ConstantRange::binaryNot() const {
   return ConstantRange(APInt::getAllOnes(getBitWidth())).sub(*this);
 }
 
+/// Estimate the 'bit-masked AND' operation's lower bound.
+///
+/// E.g., given two ranges as follows (single quotes are separators and
+/// have no meaning here),
+///
+///   LHS = [10'00101'1,  ; LLo
+///          10'10000'0]  ; LHi
+///   RHS = [10'11111'0,  ; RLo
+///          10'11111'1]  ; RHi
+///
+/// we know that the higher 2 bits of the result is always 10; and we also
+/// notice that RHS[1:6] are always 1, so the result[1:6] cannot be less than
+/// LHS[1:6] (i.e., 00101). Thus, the lower bound is 10'00101'0.
+///
+/// The algorithm is as follows,
+/// 1. we first calculate a mask to find the higher common bits by
+///       Mask = ~((LLo ^ LHi) | (LLo ^ LHi) | (LLo ^ RLo));
+///       Mask = clear all non-leading-ones bits in Mask;
+///    in the example, the Mask is set to 11'00000'0;
+/// 2. calculate a new mask by setting all common leading bits to 1 in RHS, and
+///    keeping the longest leading ones (i.e., 11'11111'0 in the example).
+/// 3. return (LLo & new mask) as the lower bound;
+/// 4. repeat the step 2 and 3 with LHS and RHS swapped, and update the lower
+///    bound with the larger one.
+static APInt estimateBitMaskedAndLowerBound(const ConstantRange &LHS,
```

Comment:
I happened to analyze the problem in Oct and made a comment to
https://stackoverflow.com/questions/2620388/bitwise-interval-arithmetic

Do the two algorithms return the same result? If yes, my code seems simpler?

Reply:
> I happened to analyze the problem in Oct and made a comment to https://stackoverflow.com/questions/2620388/bitwise-interval-arithmetic
> 
> Do the two algorithms return the same result? If yes, my code seems simpler?

Aha, I saw your answer on SO before, but I don't think they are same. E.g., given

```
a.lo = 0b110101;
a.hi = 0b111000;  // inclusive
b.lo = 0b111111;
b.hi = 0b111111;
```

your algorithm gives the lower bound as `0b110000`, while mine gives `0b110101`.